src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AECraftJob.java

```diff
@@ -0,0 +1,309 @@
+        ICraftingSubmitResult submitResult = craftingService.submitJob(job, requester, targetCpu, false, this.source);
+        if (!submitResult.successful()) {
+            calculationNotSuccessful = true;
+            fireEvent(true, submitResult.errorCode().toString());
+            return;
+        }
+
+        this.jobLink = submitResult.link();
+        this.futureJob = null;
+        setStartedCrafting();
+        prepareCPUAndStatus(craftingService);
+    }
+
+    public void jobStateChanged() {
+        ICraftingLink jobLink = this.jobLink;
+        if (jobLink == null) {
+            fireEvent(true, StatusConstants.UNKNOWN_ERROR);
+            return;
+        }
+
+        if (jobLink.isCanceled() && !isJobCanceled) {
+            fireEvent(false, StatusConstants.JOB_CANCELED);
+            setJobCanceled();
+            return;
+        }
+
+        if (jobLink.isDone() && !isJobDone) {
+            fireEvent(false, StatusConstants.JOB_DONE);
+            setJobDone();
+        }
+    }
+
+    private void prepareCPUAndStatus(ICraftingService service) {
+        if (jobLink == null || jobStatus != null || !startedCrafting) {
+            return;
+        }
+        for (ICraftingCPU cpu : service.getCpus()) {
+            if (cpu instanceof CraftingCPUCluster cpuCluster) {
+                if (cpuCluster.craftingLogic.getLastLink() != null && cpuCluster.craftingLogic.getLastLink().getCraftingID().equals(jobLink.getCraftingID())) {
+                    this.jobStatus = () -> {
+                        // Compare the id of the job in the cpu. This job object can exist longer than the job needs time to complete. So the cpu could have a new job
+                        if (cpuCluster.craftingLogic.getLastLink() != null && cpuCluster.craftingLogic.getLastLink().getCraftingID().equals(jobLink.getCraftingID()))
+                            return cpuCluster.getJobStatus();
+                        return null;
+                    };
+                    cpuCluster.craftingLogic.addListener((key) -> {
+                        // The last time the listeners are called from the cpu logic is when the job is finished
+                        // These listeners are not intended by ae2 to be used like this, but it works, and we don't modify the key
+                        if (cpuCluster.getJobStatus() != null) {
+                            this.cachedStatus = cpuCluster.getJobStatus();
+                        }
+                    });
+                    this.targetCpu = cpu;
+                    return;
+                }
+            }
+        }
+        AdvancedPeripherals.debug("Could not find CPU or job link even after job started", org.apache.logging.log4j.Level.WARN);
+    }
+
+    private CraftingJobStatus getJobStatus() {
+        if (jobStatus == null || jobStatus.get() == null && cachedStatus != null) {
+            return cachedStatus;
+        }
+        cachedStatus = jobStatus.get();
+        return jobStatus.get();
```

Comment:
```suggestion
        if (cachedStatus == null && jobStatus != null) {
            cachedStatus = jobStatus.get();
        }
        return cachedStatus;
```
I think you mean this

Reply:
That makes more sense

Reply:
Actually I'm not sure if `jobStatus.get() == null` is on purpose, like if the cache should be used only after the job is finished or it could be use all the time?

Reply:
I've added a comment which hopefully explains it a bit better